compiler/transforms/convert_linalg_to_acc.py

```diff
@@ -0,0 +1,237 @@
+        """
+        Produce a `Sequence[Operation], SSAValue` tuple for each field that contains:
+
+        - a list of operations that calculate the field value
+        - a reference to the SSAValue containing the calculated field value
+        """
+        a, b, c = op.operands
+
+        zero = arith.Constant.from_int_and_width(0, builtin.IndexType())
+        dim = memref.Dim.from_source_and_index(a, zero)
+        size = [zero, dim], dim.result
+
+        ptrs = [
+            (
+                [ptr := memref.ExtractAlignedPointerAsIndexOp.get(ref)],
+                ptr.aligned_pointer,
+            )
+            for ref in (a, b, c)
+        ]
+
+        return ptrs + [size]
+
+
+@dataclass
+class ConvertLinalgToAcceleratorPattern(RewritePattern):
+    module: builtin.ModuleOp
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: linalg.Generic, rewriter: PatternRewriter, /):
+        if op.library_call is None:
+            return
+
+        if op.library_call.data != "snax_hwpe_mult":
+            return
+
+        # grab accelerator
+        accelerator = AcceleratorConfig()
+
+        # grab arguments
+        args = accelerator.generate_vals(op)
+
+        # insert ops to calculate arguments
+        for new_ops, _ in args:
+            rewriter.insert_op_before_matched_op(new_ops)
+
+        # instantiate setup call
+        rewriter.insert_op_before_matched_op(
+            setup := acc.SetupOp(
+                [val for _, val in args], accelerator.fields, accelerator.name
+            )
+        )
+
+        # launch
+        rewriter.insert_op_before_matched_op(token := acc.LaunchOp(setup))
+
+        # await
+        rewriter.replace_matched_op(acc.AwaitOp(token))
+
+
+@dataclass
+class ConnectStatesThroughControlFlowPattern(RewritePattern):
+    walked_funcs: set[str] = field(default_factory=set)
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, func_op: func.FuncOp, rewriter: PatternRewriter, /):
+        if func_op.sym_name.data in self.walked_funcs:
+            return
+        self.walked_funcs.add(func_op.sym_name.data)
+        _walk(func_op, {}, rewriter)
+
+
+def _walk(
+    container: Region | Operation, state: dict[str, SSAValue], rewriter: PatternRewriter
+) -> dict[str, SSAValue]:
+    if isinstance(container, Operation):
+        regions = container.regions
+    else:
+        regions = [container]
+
+    for region in regions:
+        for block in region.blocks:
+            for op in block.ops:
+                # arith, memref, linalg and test ops are not relevant to
+                # accelerator setup
+                if op.name.split(".")[0] in ("arith", "memref", "linalg", "test"):
+                    continue
+                # handle acc2 dialect ops:
+                elif op.name.startswith("acc2."):
+                    # the setup op is the only relevant one for now:
+                    if isinstance(op, acc.SetupOp):
```

Comment:
then just do this instead of `elif op.op.name.startswith("acc2.")`?

Reply:
Yeah, I guess this should be cleaned up a bit further. I was somewhat unhinged when I wrote this code yesterday